core/unix/os.c

```diff
@@ -3474,10 +3474,17 @@ os_thread_suspend(thread_record_t *tr)
      */
     mutex_unlock(&ostd->suspend_lock);
     while (ksynch_get_value(&ostd->suspended) == 0) {
-        /* For Linux, waits only if the suspended flag is not set as 1. Return value
-         * doesn't matter because the flag will be re-checked.
+        /* For Linux, waits only if the suspended flag is not set as 1. Other
+         * than a timeout value, the return value doesn't matter because the
+         * flag will be re-checked.
          */
-        ksynch_wait(&ostd->suspended, 0, 0);
+        if (ksynch_wait(&ostd->suspended, 0, 5000) == -ETIMEDOUT) {
+            mutex_lock(&ostd->suspend_lock);
+            ASSERT(ostd->suspend_count > 0);
+            ostd->suspend_count--;
+            mutex_unlock(&ostd->suspend_lock);
+            return false;
```

Comment:
Hmm, I'm worried there's some use of os_thread_suspend() outside of synch_with_all_threads() (e.g., just direct synch_with_thread()) that is not prepared to retry, and that in some extreme circumstance, such as a heavily loaded machine or machine resuming from suspend-to-disk or something, where there will be a long delay for *everything* and now we'll fail and mess something up where before it worked.  Do other uses exist?  I know they do on Windows b/c of app syscalls that suspend threads.

This also splits the behavior of this cross-OS interface from Windows and should probably be mentioned in the header.

Reply:
I've added documentation about how `os_thread_suspend` may or may not respect the new `timeout_ms` parameter, and now only supply a timeout parameter to the function when the `synch_with_thread`'s `block` parameter is false. My logic is that if a caller requested that `block` is true they probably do not want the function to time-out (they expect it to "block" until the thread is synch'd), whereas if they requested a non-blocking call then a default timeout makes sense. Along with `synch_with_thread`'s `flags` parameter, the caller can get all sorts of useful functionality (wait forever, timeout -> no error, timeout -> error).

I wasn't aware of the other sorts of use-cases you mention which would cause long synch times. Does my change to infer timeouts from the `block` parameter of `synch_with_thread` make things better?

Reply:
Yes, this is better: all the other calls pass block=true.